util/scripts/sync_tas/sync_project_from_tas.py

```diff
@@ -0,0 +1,235 @@
+                    continue
+                if key == 'piId':
+                    # convert TAS user id to portal user id
+                    username = tas.get_user(id=proj[key])['username']
+                    cursor.execute("SELECT id FROM {table} WHERE username = '{username}'".format(table = PORTAL_AUTH_USER_TABLE_NAME,
+                                                                                                 username = username))
+                    portal_user = cursor.fetchone()
+                    if portal_user:
+                        val = portal_user['id']
+                    else:
+                        logger.warning('can not find user with username {} in portal'.format(username))
+                        val = None
+                reformated_proj[TAS_TO_PORTAL_MAP[key]] = val
+        # get nickname
+        if tas_proj_nickname:
+            reformated_proj['nickname'] = tas_proj_nickname
+        elif tas_proj_id:
+            cursor.execute("SELECT nickname FROM {table} WHERE tas_project_id = '{tas_project_id}'".format(table = PORTAL_PROJECT_EXTRA_TABLE_NAME,
+                                                                                                           tas_project_id = tas_proj_id))
+            nickname = cursor.fetchone()
+            if nickname:
+                reformated_proj['nickname'] = nickname['nickname']
+            else:
+                reformated_proj['nickname'] = reformated_proj['charge_code']
+                logger.warning('No nickname found in portal for project with tas id {}. The charge code will be used.'.format(tas_proj_id))
+        if is_reformated_project_missing_required_fields(reformated_proj):
+            logger.warning('TAS project is missing required fields: {}'.format(proj))
+        else:
+            result[reformated_proj['charge_code']] = reformated_proj
+    
+    logger.info('TAS has {} records'.format(len(result)))
+    
+    return result
+
+def get_projects_from_portal(db):
+    result = {}
+    cursor = db.cursor(dictionary=True)
+    cursor.execute("SELECT * FROM {}".format(PORTAL_PROJECT_TABLE_NAME))
+    for proj in cursor.fetchall():
+        result[proj['charge_code']] = proj
+        
+    logger.info('Portal has {} records'.format(len(result)))
+        
+    return result
+
+def sync(db, tas_projects, portal_projects):
+    records_to_insert = []
+    records_to_update = []
+    columns = None
+    
+    for proj_charge_code in tas_projects.keys():
+        if not columns: columns = tas_projects[proj_charge_code].keys()
+        if proj_charge_code in portal_projects:
+            portal_proj_id = portal_projects[proj_charge_code]['id']
+            del portal_projects[proj_charge_code]['id']
+            if tas_projects[proj_charge_code] != portal_projects[proj_charge_code]:
+                proj_values = []
+                for key in tas_projects[proj_charge_code].keys():
+                    proj_values.append(key)
```

Comment:
I don't understand, both the key and the value are inserted?

Reply:
that's for update query. "update <table> set <column> = <value>"